### DISCUSSION TITLE
NodeBipartiteObs features and action_set

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ml4co-competition#59] [Q&A] [answered]
Hi,

In the case of branching, action_set returns a list of valid actions for the next step in the environment, but how exactly is a 'valid action' determined? Can which actions are valid be determined from the 19 column and 5 row features for each variable node in the observation returned by `NodeBipartiteObs` to let the agent know which variables are valid and which aren't before it generates logit values? 

Additionally, would it be possible to please have a little more detail on what the following features mean:

- Column (variable) features: normed_reduced_cost, solution_frac, scaled_age
- Row (constraint) features: Bias, is_tight, scaled_age

**Comment by AntoinePrv** [comment, marked as answer]
Hi @cwfparsonson

The valid actions are determined by SCIP's [`SCIPgetLPBranchCands`](https://www.scipopt.org/doc-7.0.3/html/group__PublicBranchingMethods.php#gab19f03a5d02f7dc16db722c862651933) by default, or [`SCIPgetPseudoBranchCands`](https://www.scipopt.org/doc-7.0.3/html/group__PublicBranchingMethods.php#ga5feaeaa588ad1d8073eff66d4dbbc0c0) if you pass `pseudo=True` to the `Branching` environment constructor.

You are right that the interaction between the `action_set` and the observation function is not well documented.
You can use the `action_set` to directly index `NodeBipartiteObs.column_features`, as well as the columns of the sparse matrix `NodeBipartiteObs.edge_features`.
In NumPy terms, `obs.column_features[action_set]` are the columns in the observation that are valid branching variables / actions.
If you decide to branch on `obs.column_features[i]` for some `i`, then you can `env.step(i)`.

For the features, you mentioned, perhaps looking in the [code](https://github.com/ds4dm/ecole/blob/6cc08bb2a8dfff1d15b3e907e67a6ac22f6d9ee9/libecole/src/observation/nodebipartite.cpp#L136) would make for a better explanation. Feel free to reply if this is not clear enough.

**Reply by cwfparsonson** [reply]
> Hi @cwfparsonson
> 
> The valid actions are determined by SCIP's [`SCIPgetLPBranchCands`](https://www.scipopt.org/doc-7.0.3/html/group__PublicBranchingMethods.php#gab19f03a5d02f7dc16db722c862651933) by default, or [`SCIPgetPseudoBranchCands`](https://www.scipopt.org/doc-7.0.3/html/group__PublicBranchingMethods.php#ga5feaeaa588ad1d8073eff66d4dbbc0c0) if you pass `pseudo=True` to the `Branching` environment constructor.
> 
> You are right that the interaction between the `action_set` and the observation function is not well documented.
> You can use the `action_set` to directly index `NodeBipartiteObs.column_features`, as well as the columns of the sparse matrix `NodeBipartiteObs.edge_features`.
> In NumPy terms, `obs.column_features[action_set]` are the columns in the observation that are valid branching variables / actions.
> If you decide to branch on `obs.column_features[i]` for some `i`, then you can `env.step(i)`.
> 
> For the features, you mentioned, perhaps looking in the [code](https://github.com/ds4dm/ecole/blob/6cc08bb2a8dfff1d15b3e907e67a6ac22f6d9ee9/libecole/src/observation/nodebipartite.cpp#L136) would make for a better explanation. Feel free to reply if this is not clear enough.

Hi @AntoinePrv, many thanks for the reply. Am I correct in thinking then that SCIPgetLPBranchCands will return the fractional variables which can be branched on, which are any variables in `NodeBipartiteObs` with feature values `is_type_binary=False`, `is_type_integer=False`, and `is_type_continuous=True` (i.e. variables which are not integers)? Or do these feature values refer to integrality constraints, in which case SCIPgetLPBranchCands would return the variables which have `is_type_continuous=False` but whose current values are not yet integers?

I understand that we can use `action_set` to index the valid actions, but I'm more trying to work out if the `NodeBipartiteObs` state we are giving to the network before it generates its logits provides sufficient information for the network to know which actions are valid and which are not _before_ we filter the actions.

Thanks for linking the code, I'll have a look at the pyscipopt documentation and see if I can work out what the functions are returning.

**Reply by AntoinePrv** [reply]
> Am I correct in thinking then that SCIPgetLPBranchCands will return the fractional variables which can be branched on, which are any variables in NodeBipartiteObs with feature values is_type_binary=False, is_type_integer=False, and is_type_continuous=True (i.e. variables which are not integers)?

On the contrary, they should be a subset of integer variables.
The variable (in the problem) is integer, but it's LP (continuous) relaxation is not, which is why you can branch on it.

> Or do these feature values refer to integrality constraints, in which case SCIPgetLPBranchCands would return the variables which have is_type_continuous=False but whose current values are not yet integers?

More like this, yes. In the case where `pseudo=True` is passed to the `Branching` environment constructor, pseudo branching candidates are used, in which case they can contain continuous variables.

> I'm more trying to work out if the NodeBipartiteObs state we are giving to the network before it generates its logits provides sufficient information for the network to know which actions are valid and which are not before we filter the actions.

That's an interesting question. I believe most information should be there although in practice it may be hard for the NN to learn.
You can think of the `action_set` as an extension of the observation, so you can use it at inference time, even concatenate it with your observation if you like.
Unless you are doing something more advanced, I see no reasons *not* to use the action set to filter out the output of your NN.

**Reply by cwfparsonson** [reply]
> On the contrary, they should be a subset of integer variables.
> The variable (in the problem) is integer, but it's LP (continuous) relaxation is not, which is why you can branch on it.
> More like this, yes. In the case where `pseudo=True` is passed to the `Branching` environment constructor, pseudo branching candidates are used, in which case they can contain continuous variables.

Many thanks for clarifying!


> That's an interesting question. I believe most information should be there although in practice it may be hard for the NN to learn.
> You can think of the `action_set` as an extension of the observation, so you can use it at inference time, even concatenate it with your observation if you like.
> Unless you are doing something more advanced, I see no reasons _not_ to use the action set to filter out the output of your NN.

The reason why I thought just filtering the actions after generating logits might not be good is because for e.g. a 500x1000 set cover problem, you have ncols=1000 variables, but only ~80 of these actions tend to be in `action_set`. Therefore we'd be needlessly valuing variables which cannot be branched on. However, since looking through the code you linked, it seems that the `solution_frac` feature should be sufficient to tell the agent which variables will be in `action_set`, since non-fractional and continuous type variables (i.e. variables which cannot be branched on) will have have `solution_frac=0`, and valid actions will be >0.

The one feature I am still not sure I understand is `normed_reduced_cost`. It is the 'reduced cost of the variable in the last (feasible) LP' divided by the 'Euclidean norm of the objective function vector'. What is the meaning of the reduced cost of an individual variable? Is this a metric of how much this particular variable contributed to the cost reduction in the last LP divided by the overall objective function value as measured by the Euclidean norm?

**Reply by AntoinePrv** [reply]
> What is the meaning of the reduced cost of an individual variable?

The [reduced cost](https://en.wikipedia.org/wiki/Reduced_cost) is a notion coming from the [Simplex algorithm](https://en.wikipedia.org/wiki/Simplex_algorithm), the algorithm used by SCIP to compute the [LP relaxation](https://en.wikipedia.org/wiki/Linear_programming_relaxation) (aka the _continuous_ relaxation) on every node.
In the Simplex, you select (at each Simplex iteration, also called _pivot_) some variables to form a basis of the variable space, and express other variables in that basis. The reduced cost for a variable is the coefficient of the objective function expressed in that basis, and 0 for non-basis variables.
The basis change at each iteration, so what SCIP tells you is the the reduced cost on the final basis, when the simplex algorithm has terminated.

The normalization is done for the ML data to be bounded.
